content/en/blog/_posts/2021-07-15-server-side-apply-ga.md

```diff
@@ -0,0 +1,176 @@
+
+```go
+hpa := v1.HorizontalPodAutoscaler{
+         TypeMeta: metav1.TypeMeta{
+                  APIVersion: "autoscaling/v1",
+                  Kind:       "HorizontalPodAutoscaler",
+         },
+         ObjectMeta: ObjectMeta{
+                  Namespace: ns,
+                  Name:      autoscalerName,
+         },
+         Spec: v1.HorizontalPodAutoscalerSpec{
+                  MinReplicas: pointer.Int32Ptr(0),
+         },
+}
+```
+
+The above code attempts to declare the same apply configuration as shown in the previous examples, but when marshalled to YAML, produces:
+
+```yaml
+kind: HorizontalPodAutoscaler
+apiVersion: autoscaling/v1
+metadata
+  name: myHPA
+  namespace: myNamespace
+  creationTimestamp: null
+spec:
+  scaleTargetRef:
+    kind: ""
+    name: ""
+  minReplicas: 0
+  maxReplicas: 0
+```
+
+Which, among other things, contains `spec.maxReplicas` set to `0`. This is almost certainly not what the caller intended (the intended apply configuration says nothing about the `maxReplicas` field), and could have serious consequences on a production system: it directs the autoscaler to downscale to zero pods. The problem here originates from the fact that the go structs contain required fields that are zero valued if not set explicitly. The go structs work as intended for create and update operations, but are fundamentally incompatible with apply, which is why we have introduced the generated "apply configuration" types.
+
+The "apply configurations" also have convenience `With<FieldName>` functions that make it easier to build apply requests. This allows developers to set fields without having to deal with the fact that all the fields in the "apply configuration" types are pointers, and are inconvenient to set using go. For example `MinReplicas: &0` is not legal go code, so without the `With` functions, developers would work around this problem by using a library, .e.g. `MinReplicas: pointer.Int32Ptr(0)`, but string enumerations like `corev1.Protocol` are still a problem since they cannot be supported by a general purpose library. In addition to the convenience, the `With` functions also isolate developers from the underlying representation, which makes it safer for the underlying representation to be changed to support additional features in the future.
+
+## Using Server-side Apply in a controller
+
+The new client-go support makes it much easier to use Server-side Apply in controllers.
+
+When authoring new controllers to use Server-side Apply, a good approach is to have the controller recreate the apply configuration for an object each time it reconciles that object.  This ensures that the controller fully reconciles all the fields that it is responsible for. Controllers typically should unconditionally set all the fields they own by setting `Force: true` in the `ApplyOptions`. Controllers must also provide a `FieldManager` name that is unique to the reconciliation loop that apply is called from.
```

Comment:
I'm guessing this replaces deepEquals logic? Wouldn't this add additional load on the api server? But agree that this definitely simplifies the reconciler.

Reply:
Can you elaborate on how this replaces `deepEquals`?

Reply:
@apelisse , sorry about the confusion. I was asking if this ('recreate the apply configuration for an object each time it reconciles') phrase meant SSA is a replacement for performing `deepEquals` before sending create/update request to k8s api server to avoid infinite reconcile loop?

Reply:
Your assumptions are all reasonable and great. This should, in general, replace the `deepEqual` since controller shouldn't even have to download the object in the first place. We've spent a lot of time benchmarking and optimizating the apiserver so that this doesn't become a problem, thank you!

Reply:
Thank you!